docs/communicate-with-javascript.md

```diff
@@ -1669,3 +1670,485 @@ nullable]` above. Currently 4 directives are supported: `null_to_opt`,
+optional field in this case.
+
+##### Compatibility with OCaml features
+
+The `@bs.deriving abstract` attribute and its lightweight variant can be used
+with [mutable
+fields](https://v2.ocaml.org/manual/coreexamples.html#s:imperative-features) and
+[private types](https://v2.ocaml.org/manual/privatetypes.html), which are
+features inherited by Melange from OCaml.
+
+When the record type has mutable fields, Melange will generate setter functions
+for them. For example:
+
+```ocaml
+type person = {
+  name : string;
+  mutable age : int;
+}
+[@@bs.deriving abstract]
+
+let alice = person ~name:"Alice" ~age:20 
+
+let () = ageSet alice 21
+```
+
+This will generate:
+
+```javascript
+var alice = {
+  name: "Alice",
+  age: 20
+};
+
+alice.age = 21;
+```
+
+If the `mutable` keyword is omitted from the interface file, Melange will not
+include the setter function in the module signature, preventing other modules
+from mutating any values from the type.
+
+Private types can be used to prevent Melange from creating the constructor
+function. For example, if we define `person` type as private:
+
+```ocaml
+type person = private {
+  name : string;
+  age : int;
+}
+[@@bs.deriving abstract]
+```
+
+The accessors `nameGet` and `ageGet` will still be generated, but not the
+constructor `person`. This is useful when binding to JavaScript objects while
+preventing any Melange code from creating values of such type.
+
+### Create JavaScript objects using external functions
+
+We have already explored one approach to creating JavaScript object literals by
+using [`Js.t` values and the `bs.obj` extension](#using-jst-objects).
+
+In addition to that approach, Melange offers the `bs.obj` attribute, which can
+be used in combination with external functions to create JavaScript objects.
+When these functions are called, they generate objects with fields corresponding
+to the labeled arguments of the function.
+
+If any of these labeled arguments are defined as optional and omitted during
+function application, the resulting JavaScript object will exclude the
+corresponding fields. This feature allows for runtime object creation with
+dynamic selection of the subset of fields to include.
+
+For example, assuming we need to bind to a JavaScript object like this:
+
+```js
+var homeRoute = {
+  type: "GET",
+  path: "/",
+  action: () => console.log("Home"),
+  // options: ...
+};
+```
+
+The first three fields are required and the `options` field is optional. You can
+declare a binding function like:
+
+```ocaml
+external route :
+  _type:string ->
+  path:string ->
+  action:(string list -> unit) ->
+  ?options:< .. > ->
+  unit ->
+  _ = ""
+  [@@bs.obj]
+```
+
+Note that the empty string at the end of the function is used to make it
+syntactically valid. The value of this string is ignored by the compiler.
+
+Since there is an optional argument `options`, an additional unlabeled argument
+of type `unit` is required. This allows the compiler to determine if the
```

Comment:
The argument isn't really required, it's just a convenience that allows not having to apply every optional argument with `?foo:None`.

Feel free to leave it as is, though? it's a good recommendation, it's just not "required".

Reply:
Tried to rewrite in https://github.com/melange-re/documentation-site/pull/28/commits/04ef00465e6834d0d4bf6fbb990af8e25c51ef44.